tests/testthat/test_preprocess_ww_data.R

```diff
@@ -0,0 +1,279 @@
+test_that("lab_site_index and site_index are created correctly", {
+  processed <- preprocess_ww_data(ww_data,
+    conc_col_name = "conc",
+    lod_col_name = "lod"
+  )
+
+  expect_true(all(!is.na(processed$lab_site_index)))
+  expect_true(all(!is.na(processed$site_index)))
+
+  # Check for unique indices for each lab-site combination
+  expect_equal(length(unique(processed$lab_site_index)), 2)
+
+  # Check for unique indices for each site
+  expect_equal(length(unique(processed$site_index)), 2)
+})
+
+# Test that lab_site_name is constructed properly
+test_that("lab_site_name is constructed properly", {
+  processed <- preprocess_ww_data(ww_data,
+    conc_col_name = "conc",
+    lod_col_name = "lod"
+  )
+
+  expected_lab_site_names <- c(
+    "Site: 1, Lab: 1", "Site: 1, Lab: 1",
+    "Site: 2, Lab: 1", "Site: 2, Lab: 1"
+  )
+
+  expect_equal(processed$lab_site_name, expected_lab_site_names)
+})
+
+# Test that the function handles empty dataframes correctly
+test_that("Function handles empty dataframes with an error", {
+  empty_ww_data <- ww_data[FALSE, ]
+
+  expect_error(preprocess_ww_data(empty_ww_data,
+    conc_col_name = "conc",
+    lod_col_name = "lod"
+  ))
+})
+
+# Test that the function flags outliers correctly
+# (assuming flag_ww_outliers works as expected)
+test_that("Function flags outliers correctly", {
+  # Add a potential outlier value for testing
+  outlier_ww_data <- tibble::tibble(
+    date = seq(
+      from = lubridate::ymd("2023-11-01"),
+      to = lubridate::ymd("2023-11-20"),
+      by = "days"
+    ),
+    site = rep(1, 20),
+    lab = rep(1, 20),
+    conc = c(
+      345.2, 335.3, 345.2, 335.3, 345.2, 335.3,
+      345.2, 335.3, 334,
+      345.2, 335.3, 345.2, 335.3, 345.2, 335.3,
+      345.2, 335.3, 340, 334, 100000
+    ),
+    lod = rep(20, 20),
+    site_pop = rep(1e6, 20)
+  )
+
+  processed <- preprocess_ww_data(outlier_ww_data,
+    conc_col_name = "conc",
+    lod_col_name = "lod"
+  )
+
+  # Assuming flag_as_ww_outlier is binary (0 or 1) and
+  # the last entry is an outlier
+  expect_true(all(processed$flag_as_ww_outlier %in% c(0, 1)))
+
+  # Check if the last entry is flagged as an outlier
+  expect_equal(tail(processed$flag_as_ww_outlier, n = 1), c(1))
+})
+
+# Test that all rows are preserved after preprocessing
+test_that("All rows are preserved after preprocessing", {
+  processed <- preprocess_ww_data(ww_data,
+    conc_col_name = "conc",
+    lod_col_name = "lod"
+  )
+
+  # The number of rows should remain unchanged after processing
+  expect_equal(nrow(processed), nrow(ww_data))
+})
+
+# Test that no NA values are introduced during preprocessing
+# (assuming input has no NAs)
+test_that("No NA values are introduced during preprocessing", {
+  processed <- preprocess_ww_data(ww_data,
+    conc_col_name = "conc",
+    lod_col_name = "lod"
+  )
+
+  # Check for any new NA values introduced in any column
+  expect_true(all(complete.cases(processed)))
```

Comment:
What the reason for `complete.cases` over `!is.na`? Open to it but would like to know.

Reply:
No reason, I think this is one chat gpt suggested for me :)